docs/proposals/clusterconfig.md

```diff
@@ -0,0 +1,110 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Overview
+
+Proposes a mechanism for specifying cluster-wide runtime config
+for all kube system components via a single config file. This will
+provide an easy way to specify init settings that are shared between
+system components.
+
+## Motivation
+
+Primary motivation is enabling/disabling features that are not tied to
+an API group. API groups can be selectively enabled/disabled in the
+apiserver via the `--runtime-config` commandline flag, but there is
+currently no mechanism to toggle alpha features that are controlled by
+e.g. annotations. This means the burden of controlling whether such
+features are enabled in a particular cluster is on feature implementors;
+they must either define some ad hoc mechanism for toggling (e.g. flag
+on component binary) or else toggle the feature on/off at compile time.
+
+Providing a mechanism for specifying cluster-wide config makes
+deploying non-API features simpler for implementors and cluster admins.
+
+## Design
+
+Cluster-wide runtime config will be specified in a single yaml file
+loaded from a known, default path. TBD whether default config path
+should be overwriteable via flag/env.
+
+Strawman:
+```shell
+$ cat /etc/srv/kubernetes/cluster_config.yaml
+runtimeConfig:
+  enableAppArmor: true
+  enableFooBarBaz: false
```

Comment:
How would this work for third party extension apis (like federation, or an API group not hosted on the master)?  

To take the exact opposite direction here - shouldn't component config support a set of flags that indicate alpha feature enablement but are instead typed?  I.e. instead of map[string]string:

```
type SchedulerConfig struct {
  ...
  AlphaEnableAppArmor bool
  AlphaEnableFooBarBaz bool
}
```

That at least is discoverable (clients know that AppArmor is an option, and that it's alpha) and forces you to properly identify which components are impacted by AppArmor.

I'm kind of skeptical of the generic "enable runtime option across EVERY component in the cluster" option, although I can understand why someone is tempted to enable it.  How do rolling upgrades of enabled alpha features to beta features work?  How many components on average would we add these flags to, that justifies going a bit further and having a shared enablement config?